lib/rules/no-return-await.js

```diff
@@ -23,6 +25,24 @@ module.exports = {
     },
 
     create(context) {
+
+        /**
+        * Determines whether a thrown error from this node will be caught/handled within this function rather than immediately halting
+        * this function. For example, a statement in a `try` block will always have an error handler. A statement in
+        * a `catch` block will only have an error handler if there is also a `finally` block.
+        * @param {ASTNode} node A node representing a location where an could be thrown
+        * @returns {boolean} `true` if a thrown error will be caught/handled in this function
+        */
+        function hasErrorHandler(node) {
+            if (astUtils.isFunction(node) || node.type === "Program") {
+                return false;
+            }
+            if (node.parent.type === "TryStatement" && (node === node.parent.block || node === node.parent.handler && node.parent.finalizer)) {
+                return true;
+            }
+            return hasErrorHandler(node.parent);
```

Comment:
Correct, it could also use a `while` loop. Personally I think recursion is better than a loop in cases like this, but I'm fine with changing it if you feel otherwise.